### ISSUE TITLE
Cache::get_init_cache does out-of-bounds memory access

### ISSUE THREAD

**Issue by marmistrz** [fabianvdW/FabChess#29]
The offending code is in the following lines:
https://github.com/fabianvdW/FabChess/blob/2f49c5c74dc17560d727cfaaa60468848bfb62cb/src/search/cache.rs#L32-L69

I have absolutely no idea what's the purpose of this code. In the following lines, all the threads are simultaneously writing default values of `CacheBucket` beyond the space allocated for `cache_vec`.
https://github.com/fabianvdW/FabChess/blob/2f49c5c74dc17560d727cfaaa60468848bfb62cb/src/search/cache.rs#L40-L51

If we are only going to have default cache buckets, why can't we just calculate the required length for `cache_vec` and just use `vec![CacheBucket::default(); buckets];`? This seems to work fine for me. I'll submit a PR in a while.

**Comment by fabianvdW** [comment]
First of all, thanks for contributing, this is very helpful ! :)

I know that the code you posted is very ugly, especially for Rust, but it does serve the purpose of allocating the memory in parallel if multiple threads are provided via UCI. This is relevant for events like TCEC, where huge memory is allocated (like 64GB), and a startup time of >1min is unacceptable.

>  In the following lines, all the threads are simultaneously writing default values of CacheBucket beyond the space allocated for cache_vec.

The memory is allocated, see [Vec::with_capacity](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.with_capacity) and [Capacity and Reallocation](https://doc.rust-lang.org/std/vec/struct.Vec.html#capacity-and-reallocation). The length of the vector isn't updated, but it is done with the line `cache_vec.set_len(local_len); ` later.

> If we are only going to have default cache buckets, why can't we just calculate the required length for cache_vec and just use vec![CacheBucket::default(); buckets];?
It was so earlier, but it doesn't allow for parallel initialization. That said, if you know/there are other ways to improve on this code, I would be happy to! A first step would be to change
```Rust
let mut ptr = cache_vec.as_mut_ptr().add(cache_vec.len()); 
let mut local_len = cache_vec.len(); 
```
to 
```Rust
let mut ptr = cache_vec.as_mut_ptr();
let mut local_len = 0;
```
, as cache_vec is empty in the beginning.

Fabi

**Comment by marmistrz** [comment]
I see. This piece of code definitely needs comments. I missed the fact that `cache_vec.len() == 0` at the beginning, this fact was obscured by an allocator WASI implementation bug :) I have an idea how to clean up the code, I'll update the PR soon.

**Comment by marmistrz** [comment]
I updated the PR. Ideally, we'd use something like [`par_iter_mut`](https://docs.rs/rayon/1.3.0/rayon/iter/trait.IntoParallelRefMutIterator.html#tymethod.par_iter_mut) instead, but we don't want to initialize a rayon thread pool **just** to initialize the vector.

**Comment by fabianvdW** [comment]
merged with #33
